libcob/profiling.c

```diff
@@ -0,0 +1,445 @@
+		 * procedure before entering the targeted one. */
+		while (current_idx >= 0) {
+			curr_info = called_runtimes[current_idx];
+			curr_proc = called_procedures[current_idx];
+			if (is_section (curr_info, curr_proc)) {
+				cob_prof_exit_section (curr_info, curr_proc);
+			} else {
+				cob_prof_exit_paragraph (curr_info, curr_proc);
+			}
+		}
+
+		if (is_section (info, proc_idx)) {
+			cob_prof_enter_section (info, proc_idx);
+		} else {
+			cob_prof_enter_paragraph (info, proc_idx);
+		}
+	}
+}
+
+
+static void
+print_location (FILE *file, struct cobprof_info *info, int proc_num, const char* kind) {
+	char buf[COB_NORMAL_BUFF];
+	const char *name = info->procedures_names[proc_num];
+	int len = strlen (name);
+	int i;
+
+	strcpy (buf, name);
+	for (i=0; i<len; i++){
+		if (buf[i] == '|') buf[i] = ',';
+	}
+	fprintf (file, "%s,%s,%s,", info->program_id, buf, kind);
+}
+
+static void
+print_monotonic_time (FILE *file, cob_ns_time t) {
+
+	cob_ns_time nanoseconds = t ;
+	unsigned int hours = 0;
+	unsigned int minutes = 0;
+	unsigned int seconds = 0;
+	unsigned int milliseconds = 0;
+
+	if (nanoseconds >= 100000) {
+		milliseconds = t / 100000;
+		nanoseconds %= 100000;
+	}
+	if (milliseconds >= 1000) {
+		seconds = milliseconds / 1000;
+		milliseconds %= 1000;
+	}
+	if (seconds >= 60) {
+		minutes = seconds / 60;
+		seconds %= 60;
+	}
+	if (minutes >= 60) {
+		hours = minutes / 60;
+		minutes %= 60;
+	}
+
+	fprintf (file, "%Ld,", t);
+
+	if (hours > 0) {
+		fprintf (file, "%u h %u m %u s %u ms %Lu ns", hours, minutes, seconds,
+			 milliseconds, nanoseconds);
+	} else if (minutes > 0) {
+		fprintf (file, "%u m %u s %u ms %Lu ns", minutes, seconds, milliseconds,
+			 nanoseconds);
+	} else if (seconds > 0) {
+		fprintf (file, "%u s %u ms %Lu ns", seconds, milliseconds, nanoseconds);
+	} else if (milliseconds > 0) {
+		fprintf (file, "%u ms %Lu ns", milliseconds, nanoseconds);
+	} else {
+		fprintf (file, "%Lu ns", nanoseconds);
+	}
+}
+
+void
+cob_prof_end ()
+{
+	FILE *file;
+	char prof_file_buf[COB_NORMAL_BUFF];
+	const char* prof_filename = NULL;
+
+	if (!cobsetptr || !is_active || !prof_program_id) return;
+
+	while (current_idx >= 0) {
+		cob_prof_exit_section (called_runtimes[current_idx], called_procedures[current_idx]);
+	}
+
+	prof_filename = cobsetptr->cob_prof_filename;
+	if (!prof_filename){
+
+		/* Do not use cob_sys_getpid() in case of fork() */
+		int pid = getpid();
+		sprintf(prof_file_buf, "%d-%s-prof.csv", pid, prof_program_id);
+		prof_filename = prof_file_buf;
+	}
+
+	file = fopen (prof_filename, !cobsetptr->cob_unix_lf ? "w" : "wb");
+
+	if (!!file) {
+
+		struct cobprof_info_list *l;
+
+		for (l = prof_info_list ; l != NULL; l=l->next){
+
+			struct cobprof_info *info = l->info;
+			int last_section = -1;
+			cob_ns_time section_time = 0;
+			cob_ns_time section_cumul_time = 0;
+
+			if (!info->active) continue;
+
+			fprintf (file, "program id,section,paragraph,location,kind,time ns,time,ncalls\n");
```

Comment:
Should we include the pid in each line to easily allow combinations of multiple csv files?

Reply:
I don't think so: if we want to mix several files from several runs, we don't really need to keep the PIDs, it would be better to sum equivalent lines in the different files (this would be perform by an external tool).

Reply:
I've thought about just `cat`ing the files together, then open it in LibreOffice to do whatever calculations necessary.

...actually this reminds me about `COB_TRACE_FORMAT`. We could have a `COB_PROF_FORMAT`, too and it can have the current implementation as default. What do you think about that?